core/vm/contract.go

```diff
@@ -92,6 +92,38 @@ func (c *Contract) validJumpdest(dest *big.Int) bool {
 	if OpCode(c.Code[udest]) != JUMPDEST {
 		return false
 	}
+	return c.isCode(udest)
+}
+
+func (c *Contract) validJumpSubdest(udest uint64) bool {
+	// PC cannot go beyond len(code) and certainly can't be bigger than 63 bits.
+	// Don't bother checking for BEGINSUB in that case.
+	if int64(udest) < 0 || udest >= uint64(len(c.Code)) {
+		return false
+	}
+	// Only BEGINSUBs allowed for destinations
+	if OpCode(c.Code[udest]) != BEGINSUB {
+		return false
+	}
+	return c.isCode(udest)
+}
+
+func (c *Contract) validReturndest(udest uint64) bool {
```

Comment:
Do we really need to check if something is a valid returndest? 
I mean
```golang
   pos := returns.pop()
	if !contract.validReturndest(pos) {
		return nil, errInvalidJump
	}
```
The argument `udest` is always coming from the `returns` (return stack). And nothing should ever wind up there that is not already valid (as opposed to the normal stack, which can contain anything). I think we can nuke this

Reply:
fixed